### DISCUSSION TITLE
Why use V(s) instead of Q(s, a) in computing TD error in Actor-Critic agent?

### DISCUSSION THREAD

**Discussion by QuantHao** [PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook#30] [Reinforcement Learning] [answered]
Hi, @praveen-palanisamy 

I notice that in when you teach SARSA and Q-learning in Chapter2, you always use Q(s_t, a_t) when computing TD error, only difference is SARSA use Q(s_[t+1], a_[t+1]) while Q-learning use max(Q(s_[t+1], :)).

SARSA:
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/1184e86f621033852576ccb1c2893fca064ec09e/Chapter02/5_sarsa_sarsa_lambda.py#L29-L30

Q-learning:
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/1184e86f621033852576ccb1c2893fca064ec09e/Chapter02/6_q_learning.py#L28-L31

HOWEVER, when you teach Actor-Critic, it seems that you use V(s) in stead of Q(s_t, a_t) to compute TD error
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/1184e86f621033852576ccb1c2893fca064ec09e/Chapter02/8_actor_critic_agent.py#L62-L64
AND, critic network is design to only output lenght 1 instead of length action_dim
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/1184e86f621033852576ccb1c2893fca064ec09e/Chapter02/8_actor_critic_agent.py#L12-L17

Is that any specific reason for this design?? Thanks in advance.

**Comment by praveen-palanisamy** [comment, marked as answer]
Good observation!
That's because, in the Actor-Critic agent recipe, the policy gradient is computed using the (one-step) TD-error and for this, the Critic network only needs to learn the state-value function V(s), which is a scalar (length=1).                      **(TD Actor-Critic)**

There's an equivalent policy gradient form called _Q Actor-Critic_ that uses the action-value function Q(s, a),  but then the Critic network needs to learn the action-value function Q(s, a), which would be a vector (length=action_dim). Due to the higher-dimensional nature of Q(s, a), it is a harder function to be approximated/learned by a neural network and practically it's often unstable.                 **(Q Actor-Critic)**

There's yet another equivalent policy gradient form called _Advantage Actor-Critic_ that uses the advantage function A(s, a) but then the Critic network needs to learn both V(s) and Q(s, a) since A(s, a) = Q(s, a) - V(s). This would require two sets of Critic parameters (or a shared network with two output heads) -- One for approximating/learning V(s) & another for Q(s, a).         **( Advantage Actor-Critic)**

Using the Advantage function **(Advantage Actor-Critic)** is better than using the Q function because the Advantage function `A(s, a) = Q(s, a) - V(s)` has less variance in estimating the policy gradient compared to Q(s,a) due to the subtraction of the "baseline" i.e the state-value function V(s). The good news is that, the TD error for the true value function is an unbiased estimate of the Advantage function! Therefore, for all practical purposes, we can use the (approximate) TD error to estimate the policy gradient as in
 https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/1184e86f621033852576ccb1c2893fca064ec09e/Chapter02/8_actor_critic_agent.py#L64
and this requires only one set of critic parameters for the state-value function V(s) which is a scalar (length=1) for a given state and therefore is easier to be approximated/learned by a neural network and is relatively stable (compared to learning Q-values).

Hope this helps.

**Reply by QuantHao** [reply]
Thanks so much for this detailed reply! Maybe you could add it into your book in the next published version.
